### PR TITLE
osd/PrimaryLogPG: record prior_version for DELETE events

### PR DESCRIPTION
**Pull request by liewegas**
Leaving this blank is inaccurate.  It also confuses the rewind
divergent log code, leading to behavior like

      ldpp_dout(dpp, 10) << __func__ << ": hoid " << hoid
			 << " prior_version or op type indicates creation,"
			 << " deleting"
			 << dendl;

Fixes: http://tracker.ceph.com/issues/20274
Signed-off-by: Sage Weil <sage@redhat.com>

### CODE REVIEW

**Comment by liewegas** [comment]
@jdurgin does this make sense?  i'm not too familiar with the error events.

**Comment by liewegas** [comment]
I updated _merge_object_diverte_entries to just filter out and ignore ERROR entries... I think that makes sense.

add_next_event looks okay to me since it is already checking for is_update() or is_delete().  I don't understand what the Tracker part is doing, though?

**Comment by jdurgin** [comment]
ah, I was looking at my branch. The tracker part is keeping a delta to determine what needs to be written to disk the next time the missing set is persisted. Since the error entry doesn't changed the missing set, we can skip that call for ERROR entries.

File: src/osd/PGLog.h [hunk 1]
```diff
@@ -755,21 +755,30 @@ struct PGLog : DoutPrefixProvider {
...
-      if (i != entries.begin() && i->prior_version != eversion_t()) {
-	// in increasing order of version
-	assert(i->version > last);
-	// prior_version correct
-	assert(i->prior_version == last);
+    list<pg_log_entry_t>::const_iterator i = entries.begin();
+    while (i != entries.end()) {
+      if (i->is_error()) {
+	ldpp_dout(dpp, 20) << __func__ << ": ignoring " << *i << dendl;
+	i = entries.erase(i);
```
Comment (jdurgin) [inline src/osd/PGLog.h]: isn't entries const?

File: src/osd/osd_types.h [hunk 2]
```diff
@@ -3757,8 +3757,9 @@ class pg_missing_set : public pg_missing_const_i {
     } else if (e.is_delete()) {
       rm(e.soid, e.version);
     }
-
-    tracker.changed(e.soid);
+    if (!e.is_error()) {
```
Comment (jdurgin) [inline src/osd/osd_types.h]: now that I look the only 2 callers of this already avoid adding ERROR entries, so this isn't needed

**Comment by liewegas** [comment]
this one doesn't:

      // update local version of peer's missing list!
      if (m && pi.last_backfill != hobject_t()) {
        for (list<pg_log_entry_t>::iterator p = m->log.log.begin();
             p != m->log.log.end();
             ++p)
	  if (p->soid <= pi.last_backfill)
	    pm.add_next_event(*p);
      }

i'll fix it there.

**Comment by jdurgin** [comment]
'make check' error looks real

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-06-20_16:21:45-rados-wip-sage-testing2-distro-basic-smithi/

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-06-21_22:08:07-rados-wip-sage-testing-distro-basic-smithi/

**Pull request merged**